src/diagnostics.rs

```diff
@@ -0,0 +1,461 @@
+    pub fn error(&self, kind: Box<dyn AsFail>) {
+        self.emit(MessageLevel::Error, kind)
+    }
+
+    pub fn warning_with_source_snippet<'ctx>(
+        &self,
+        spanned: Spanned<Box<dyn AsFail>>,
+        file: &AsciiFile<'ctx>,
+    ) {
+        self.emit_with_source_snippet(MessageLevel::Warning, spanned, file)
+    }
+
+    pub fn error_with_source_snippet<'ctx>(
+        &self,
+        spanned: Spanned<Box<dyn AsFail>>,
+        file: &AsciiFile<'ctx>,
+    ) {
+        self.emit_with_source_snippet(MessageLevel::Error, spanned, file)
+    }
+
+    pub fn emit_with_source_snippet<'ctx>(
+        &self,
+        level: MessageLevel,
+        spanned: Spanned<Box<dyn AsFail>>,
+        file: &AsciiFile<'ctx>,
+    ) {
+        let msg = Message {
+            level,
+            kind: spanned.data,
+        };
+
+        let mut writer = self.writer.borrow_mut();
+        msg.write_colored_with_code(&mut **writer, &spanned.span, file);
+        // TODO: store span
+        self.messages.borrow_mut().push(msg);
+    }
+}
+
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub enum MessageLevel {
+    Error,
+    Warning,
+}
+
+impl MessageLevel {
+    fn color(self) -> Option<Color> {
+        // Don't be confused by the return type. `None` means default color!
+        match self {
+            MessageLevel::Error => Some(Color::Red),
+            MessageLevel::Warning => Some(Color::Yellow),
+        }
+    }
+
+    fn name(&self) -> &str {
+        match self {
+            MessageLevel::Error => "error",
+            MessageLevel::Warning => "warning",
+        }
+    }
+}
+
+pub struct Message {
+    pub level: MessageLevel,
+    pub kind: Box<dyn AsFail>,
+}
+
+///
+/// Calls to functions should pass the raw writer, each function should
+/// create its own ColorOutput object that is dropped on return. This
+/// gurantees correct coloring in nested calls.
+struct ColorOutput<'a> {
+    writer: &'a mut dyn WriteColor,
+    spec: ColorSpec,
+}
+
+impl<'a> ColorOutput<'a> {
+    fn new(writer: &'a mut dyn WriteColor) -> Self {
+        writer.reset().ok();
+
+        Self {
+            writer,
+            spec: ColorSpec::new(),
+        }
+    }
+
+    fn set_color(&mut self, color: Option<Color>) {
+        // ignore coloring failures using ok()
+        self.spec.set_fg(color);
+        self.writer.set_color(&self.spec).ok();
+    }
+
+    fn set_bold(&mut self, yes: bool) {
+        // ignore coloring failures using ok()
+        self.spec.set_bold(yes);
+        self.writer.set_color(&self.spec).ok();
+    }
+
+    fn writer(&mut self) -> &mut dyn WriteColor {
+        self.writer
+    }
+}
+
+/// reset to no color by default. Otherwise code that
+/// is not color aware will print everything in the
+/// color last used.
+impl<'a> Drop for ColorOutput<'a> {
+    fn drop(&mut self) {
+        // ignore coloring failures using ok()
+        self.writer.reset().ok();
+    }
+}
+
+const MAX_CONTEXT_LENGTH: usize = 80;
```

Comment:
given we use `120` (?) length, and even `180` might be common, I would advocate for `180` here.

Reply:
Just to clarify: This is a soft limit (since its defined on bytes instead of number of columns required to render the bytes) and per side relativ to the error position.

Since its only there to deal with generated code with a few hundred chars per line, pumping this to 120 might be a good idea, but could result in (120*2+1)*4 columns being displayed in the worst case (for a long line of tab characters and the middle tab character being the error position)